Finish plot testing

ericpre opened the pull request:
- [x] tidy up the way the matplotlib backend is set
- [x] set the backend to 'agg' for testing on all platforms
- [x] get plot testing working on travis os x
- [x] reduce the memory used by the figure created during plot testing: close/delete the figure more properly
- [x] Fix a test failing in appveyor, because of an unncessary plt.show()
- [x] update dev guide

ericpre committed:
start Xvfb on travis osx

ericpre committed:
Set the matplotlib backend to 'agg' to run tests on all platforms.
Tidy up the way the matplotlib backend is set.

ericpre committed:
Remove skipping plot testing on mac os x

ericpre committed:
Remove setting matplotlib backend to 'agg' in samfire_worker.py
Add test to check if the image comparison are actually working

ericpre committed:
Add mpl_cleanup pytest fixture in conftest.py
Use mpl_cleanup fixture to delete figure generated by plot tests
Reduce memory usage by x5.
Remove completely @cleanup matplotlib decorator

ericpre committed:
Same game: mpl_cleanup and remove skip plot tests for osx

ericpre committed:
Remove old baseline images for nose tests

ericpre committed:
Set MPLBACKEND environment variable in appveyor
Remove unnecessary plt.show() in drawing/utils.py: appveyor was failing because of this, switching backend to Tk and waiting for a windows to close...
Improve plot testing of signal1D

ericpre committed:
Fix error with path
Tidy up setting up traits on headless system (remove duplicated code)

ericpre committed:
update plot testing section of the dev guide

ericpre commented on hyperspy/conftest.py:
```diff
@@ -1,12 +1,17 @@
+# matplotlib backend to 'agg' as early as we can will be useless.
+# However, resetting the rcParams to matplotlib default does the job.
+import matplotlib.pyplot as plt
+plt.rcParams.clear()
```
As documented, this bit resets the matplotlib backend to 'agg'. Do we want the tests to be run locally with 'agg'? Or would it better to test other backends, which will be the default in the system? The latter could be good to test more stuff but CI will not be necessarily consistent with the tests run locally and it could be a bit confusing if not aware of this detail.

francisco-dlp replied:
I think that locally we should test using the default mpl backend in the system. Otherwise [this test](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/tests/drawing/test_utils.py#L33) will always be skipped.

ericpre replied:
Yes and this tests is actually failing using the `TkAgg` on my system... :(
The error says "AssertionError: Expected 'mock' to be called once. Called 11 times."...

francisco-dlp replied:
That sounds like a bug...

ericpre commented:
I m quite happy with the status of plot testing now, there are a couple of things, which have been significantly improved in this PR.
This is ready to review and once this is merged, I will add a few others tests in another PR.

ericpre committed:
Do not set matplotlib backend to 'agg' on local install
Typo user guide

ericpre committed:
Fix bug when closing figure with Tk backend

ericpre commented on hyperspy/samfire_utils/samfire_worker.py:
```diff
@@ -23,9 +23,6 @@
-import matplotlib
-
-matplotlib.rcParams['backend'] = 'Agg'
 
```
@to266, I removed this when I tidied up the way the matplotlib backend is set. Could you please confirm that there is no particular reason to have this here?

francisco-dlp replied:
@to266 will know better, but I think that it'll be good to keep things as they were in this case because the worker will certainly never need interactive plotting and, as far as I remember, this was preventing a backend related crash.

ericpre replied:
Ok, let wait for @to266 answer!

ericpre replied:
I reverted this change.

francisco-dlp commented on hyperspy/conftest.py:
```diff
@@ -1,12 +1,14 @@
 import pytest
 import numpy as np
-import matplotlib.pyplot as plt
+import matplotlib
```
Is this needed?

ericpre replied:
What do you mean exactly?

francisco-dlp replied:
importing matplotlib

ericpre replied:
It is used in the `mpl_cleanup` pytest fixture, am i missing something obvious?!

francisco-dlp commented on hyperspy/tests/drawing/test_utils.py:
```diff
@@ -15,23 +15,15 @@
-@cleanup
 def test_create_figure():
-    dummy_warning = 'dummy_function have been called after closing windows'
     if matplotlib.get_backend() not in ("GTKAgg", "WXAgg", "TkAgg", "Qt4Agg"):
```
Maybe this is not needed anymore?

ericpre replied:
I would say, this test is still useful, although this is not tested in CI, we spotted a bug recently with this one!

francisco-dlp replied:
Sorry for not explaining myself: I was questioning if we should not let the test run for all backends since all of them must be supported now that you are using the right mpl event for closing the figure.

ericpre replied:
yes!

francisco-dlp commented on hyperspy/tests/drawing/test_mpl_testing_setup.py:
```diff
@@ -16,14 +16,35 @@
 
-def test_mlp_agg_for_testing():
-    assert matplotlib.get_backend() == 'agg'
+def test_mlp_agg_for_CI_testing():
```
Why is this test needed? If everything works with something other than the agg backend there is nothing wrong about it...

ericpre replied:
For CI, some tests will work and other will hold the test.
For example, this was the case with [one test in appveyor](https://ci.appveyor.com/project/hyperspy/hyperspy/build/1.0.2106), which was on hold, because it was waiting for a figure to be closed, or something similar! This happen because there was a `plt.show()` call in one function somewhere.
The question is then: where is the bug? Is it the `plt.show()` call, or is it the test?
By having this test, I was thinking, that it would be easier to find out what is wrong.

francisco-dlp reviewed (commented):
Just some small comments.

ericpre committed:
Re-introduce 'matplotlib.use('agg') in samfire_worker
Skip test testing if image comparison is working when --mpl is not called.
test_create_figure should work for all matplotlib backend except 'agg'

to266 commented on hyperspy/tests/drawing/test_mpl_testing_setup.py:
```diff
@@ -16,14 +16,40 @@
+                    reason="'mpl' plugin not call.")
+@pytest.mark.xfail(reason="Check if plotting tests are working: if this failed,"
+                   "it means that the image comparison of the plotting test are"
+                   " not working.",
```
Here the `reason` and commented explanations do not match. I assume it's just the wording of the reason that could be improved?

Also, `skipif` reason should have `"...call"` -> `"...called"`

ericpre replied:
yes, the reason is wrong...

ericpre committed:
fix typo in test_mpl_testing_setup.py 

thanks to @to266.

thomasaarholt commented:
At least one other PR (#1382) relies on this. I'm unsure how to test this, so could someone else take a look?
